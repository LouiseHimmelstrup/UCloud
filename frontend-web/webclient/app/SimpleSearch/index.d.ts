import { match } from "react-router-dom";
import PromiseKeeper from "PromiseKeeper";
import { Page } from "Types";
import { Application } from "Applications";
import { File } from "Files";
import { ProjectMetadata } from "Metadata/api";
import { History } from "history";
import { Dispatch } from "redux";
import { HeaderSearchType } from "DefaultObjects";

export interface SimpleSearchProps extends SimpleSearchOperations, SimpleSearchStateProps {
    match: match<{ 0: string, priority: string }>
    history: History
}

export interface SimpleSearchStateProps {
    files: Page<File>
    filesLoading: boolean
    applications: Page<Application>
    applicationsLoading: boolean
    projects: Page<ProjectMetadata>
    projectsLoading: boolean
    error?: string
    search: string
}

export interface SimpleSearchOperations {
    setFilesLoading: (loading: boolean) => void
    setApplicationsLoading: (loading: boolean) => void
    setProjectsLoading: (loading: boolean) => void
    setError: (error?: string) => void
    searchFiles: (query: string, page: number, itemsPerPage: number) => void
    searchApplications: (query: string, page: number, itemsPerPage: number) => void
    searchProjects: (query: string, page: number, itemsPerPage: number) => void
    setFilesPage: (page: Page<File>) => void
    setApplicationsPage: (page: Page<Application>) => void
    setProjectsPage: (page: Page<ProjectMetadata>) => void
    setSearch: (search: string) => void
    setPrioritizedSearch: (st: HeaderSearchType) => void
}