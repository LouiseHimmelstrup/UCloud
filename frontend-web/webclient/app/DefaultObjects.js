import {Cloud} from "../authentication/SDUCloudObject"
import {tusConfig} from "./Configurations";
import Uppy from "uppy";

export const DefaultStatus = {
    title: "No Issues",
    level: "NO ISSUES",
    body: "The system is running as intended.",
};

export const RightsMap = {
    NONE: 0,
    READ: 1,
    READ_WRITE: 2,
    OWN: 3,
};

export const AnalysesStatusMap = {
    "PENDING": 0,
    "IN PROGRESS": 1,
    "COMPLETED": 2,
};

export const RightsNameMap = {
    NONE: "None",
    READ: "Read",
    READ_WRITE: "Read/Write",
    OWN: "Own",
};

export const SensitivityLevel = {
    "OPEN_ACCESS": "Open Access",
    "CONFIDENTIAL": "Confidential",
    "SENSITIVE": "Sensitive",
};

export const SensitivityLevelMap = {
    "OPEN_ACCESS": 0,
    "CONFIDENTIAL": 1,
    "SENSITIVE": 2,
};

const initializeUppy = (restrictions) => 
    Uppy.Core({
        autoProceed: false,
        debug: false,
        restrictions: restrictions,
        meta: {
            sensitive: false,
        },
        onBeforeUpload: () => {
            return Cloud.receiveAccessTokenOrRefreshIt().then((data) => {
                tusConfig.headers["Authorization"] = `Bearer ${data}`;
            });
        }
    }).use(Uppy.Tus, tusConfig);

export const initObject = {
    files: {
        files: [],
        filesShown: 10,
        filesPerPage: 10,
        currentFilesPage: 0,
        loading: false,
        path: "",
        filesInfoPath: "",
        filesInfoFile: null,
        projects: []
    },
    uppy: {
        uppyFiles: initializeUppy({maxNumberOfFiles: false}),
        uppyFilesOpen: false,
        uppyRunApp: initializeUppy({ maxNumberOfFiles: 1}),
        uppyRunAppOpen: false
    },
    status: {
        status: DefaultStatus,
        title: ""
    }
};