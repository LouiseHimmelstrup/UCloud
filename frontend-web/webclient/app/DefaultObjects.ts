import {SidebarOption, Page} from "Types";
import {Status} from "Navigation";
import {Analysis, DetailedApplicationSearchReduxState, RunsSortBy} from "Applications";
import {File, DetailedFileSearchReduxState} from "Files";
import {SortOrder, SortBy} from "Files";
import {DashboardStateProps} from "Dashboard";
import {Publication} from "Zenodo";
import {Notification} from "Notifications";
import {Upload} from "Uploader";
import {Activity, ActivityGroup, ActivityFilter} from "Activity";
import {Reducer} from "redux";
import {SimpleSearchStateProps} from "Search";
import * as ApplicationRedux from "Applications/Redux";
import * as AccountingRedux from "Accounting/Redux";
import * as SnackbarRedux from "Snackbar/Redux";
import * as FavoritesRedux from "Favorites/Redux";
import {defaultAvatar} from "UserSettings/Avataaar";
import {SidebarPages} from "ui-components/Sidebar";
import {ScrollResult} from "Scroll/Types";
import * as ProjectRedux from "Project/Redux";

export const DefaultStatus: Status = {
    title: "No Issues",
    level: "NO ISSUES",
    body: "The system is running as intended."
};

export enum KeyCode {
    ENTER = 13,
    ESC = 27
}

export const emptyPage: Page<any> = {items: [], itemsPerPage: 25, itemsInTotal: 0, pageNumber: 0, pagesInTotal: 0};

export enum AnalysesStatusMap {
    "PENDING",
    "IN PROGRESS",
    "COMPLETED"
}

export enum RightsNameMap {
    "NONE" = "None",
    "READ" = "Read",
    "READ_WRITE" = "Read/Write"
}

export enum SensitivityLevel {
    "INHERIT" = "Inherit",
    "PRIVATE" = "Private",
    "CONFIDENTIAL" = "Confidential",
    "SENSITIVE" = "Sensitive"
}

export type Sensitivity = keyof typeof SensitivityLevel;

export enum SensitivityLevelMap {
    INHERIT = "INHERIT",
    PRIVATE = "PRIVATE",
    CONFIDENTIAL = "CONFIDENTIAL",
    SENSITIVE = "SENSITIVE"
}

function getFilesSortingColumnOrDefault(columnIndex: 0 | 1): SortBy {
    const sortingColumn = window.localStorage.getItem(`filesSorting${columnIndex}`);
    if (sortingColumn && Object.values(SortBy).includes(sortingColumn)) return sortingColumn as SortBy;
    switch (columnIndex) {
        case 0:
            window.localStorage.setItem("filesSorting0", SortBy.MODIFIED_AT);
            return SortBy.MODIFIED_AT;
        case 1:
            window.localStorage.setItem("filesSorting1", SortBy.SIZE);
            return SortBy.SIZE;
    }
}

function getItemOrDefault<T, T2>(itemName: string, defaultValue: T, en: T2): T {
    const item = window.localStorage.getItem(itemName);
    if (item && Object.values(en).includes(item)) {
        return item as unknown as T;
    }
    return defaultValue;
}

export interface ComponentWithLoadingState {
    loading: boolean
    error?: string
}

export interface ComponentWithPage<T> extends ComponentWithLoadingState {
    page: Page<T>
}

export interface ComponentWithScroll<Item, OffsetType> extends ComponentWithLoadingState {
    scroll?: ScrollResult<Item, OffsetType>
}

export interface ResponsiveReduxObject {
    mediaType: string
    orientation: string
    lessThan: Record<string, boolean>
    greaterThan: Record<string, boolean>
    is: Record<string, boolean>
}

export const initResponsive = (): ResponsiveReduxObject => ({
    mediaType: "",
    orientation: "",
    lessThan: {},
    greaterThan: {},
    is: {},
});

export interface FilePreviewReduxState {
    file?: File
    error?: string
}

export interface FilesReduxObject extends ComponentWithPage<File> {
    sortOrder: SortOrder
    sortBy: SortBy
    path: string
    filesInfoPath: string
    fileSelectorError?: string
    sortingColumns: [SortBy, SortBy]
    fileSelectorLoading: boolean
    fileSelectorShown: boolean
    fileSelectorPage: Page<File>
    fileSelectorPath: string
    fileSelectorCallback: Function
    fileSelectorIsFavorites: boolean
    disallowedPaths: string[]
    invalidPath: boolean
}

export interface FileInfoReduxObject {
    file?: File
    error?: string
    activity: Page<Activity>
    loading: boolean
}

export interface AnalysisReduxObject extends ComponentWithPage<Analysis> {
    sortBy: RunsSortBy
    sortOrder: SortOrder
}

export interface NotificationsReduxObject {
    redirectTo: string
    items: Notification[]
    loading: boolean
    error?: string
}

export interface ZenodoReduxObject extends ComponentWithPage<Publication> {
    connected: boolean
}

export interface StatusReduxObject {
    status: Status
    title: string
    page: SidebarPages
    loading: boolean
}

export interface SidebarReduxObject {
    pp: boolean
    options: SidebarOption[]
    kcCount: number
}

export interface HeaderSearchReduxObject {
    prioritizedSearch: HeaderSearchType
    refresh?: () => void
}

export interface RunApplicationReduxObject {

}

export type ActivityReduxObject = ComponentWithScroll<ActivityGroup, number> & ActivityFilter;

export type HeaderSearchType = "files" | "applications" | "projects";

export interface UploaderReduxObject {
    uploads: Upload[]
    visible: boolean
    path: string
    allowMultiple: boolean
    onFilesUploaded: (p: string) => void
    error?: string
    loading: boolean
}

interface LegacyReducers {
    dashboard?: Reducer<DashboardStateProps>
    files?: Reducer<FilesReduxObject>
    uploader?: Reducer<UploaderReduxObject>
    status?: Reducer<StatusReduxObject>
    notifications?: Reducer<NotificationsReduxObject>
    analyses?: Reducer<AnalysisReduxObject>
    zenodo?: Reducer<ZenodoReduxObject>
    header?: Reducer<HeaderSearchReduxObject>
    sidebar?: Reducer<SidebarReduxObject>
    activity?: Reducer<ActivityReduxObject>
    detailedResult?: Reducer<DetailedResultReduxObject>
}

export type Reducers = LegacyReducers & ApplicationRedux.Reducers & AccountingRedux.Reducers;

export type DetailedResultReduxObject = ComponentWithPage<File>

export const initDetailedResult = (): DetailedResultReduxObject => ({
    page: emptyPage,
    loading: false,
    error: undefined
});

/* FIXME */
interface LegacyReduxObject {
    dashboard: DashboardStateProps
    files: FilesReduxObject,
    uploader: UploaderReduxObject
    status: StatusReduxObject,
    notifications: NotificationsReduxObject
    analyses: AnalysisReduxObject
    zenodo: ZenodoReduxObject
    header: HeaderSearchReduxObject
    sidebar: SidebarReduxObject
    activity: ActivityReduxObject
    detailedResult: DetailedResultReduxObject
    simpleSearch: SimpleSearchStateProps
    detailedFileSearch: DetailedFileSearchReduxState
    detailedApplicationSearch: DetailedApplicationSearchReduxState
    fileInfo: FileInfoReduxObject
    avatar: AvatarReduxObject
    filePreview: FilePreviewReduxState
    responsive?: ResponsiveReduxObject
    project: ProjectRedux.State
    loading?: boolean
}

export type ReduxObject =
    LegacyReduxObject &
    ApplicationRedux.Objects &
    AccountingRedux.Objects &
    FavoritesRedux.Objects &
    SnackbarRedux.Wrapper;

export const initActivity = (): ActivityReduxObject => ({
    loading: false
});

export const initNotifications = (): NotificationsReduxObject => ({
    items: [],
    loading: false,
    redirectTo: "",
    error: undefined
});

export const initHeader = (): HeaderSearchReduxObject => ({
    prioritizedSearch: "files"
});

export const initStatus = (): StatusReduxObject => ({
    status: DefaultStatus,
    title: "",
    page: SidebarPages.None,
    loading: false
});

export const initDashboard = (): DashboardStateProps => ({
    favoriteFiles: [],
    recentFiles: [],
    recentAnalyses: [],
    notifications: [],
    favoriteLoading: false,
    recentLoading: false,
    analysesLoading: false
});

export function initObject(homeFolder: string): ReduxObject {
    return {
        dashboard: initDashboard(),
        files: initFiles(homeFolder),
        status: initStatus(),
        header: initHeader(),
        notifications: initNotifications(),
        analyses: initAnalyses(),
        zenodo: initZenodo(),
        sidebar: initSidebar(),
        uploader: initUploads(),
        activity: initActivity(),
        detailedResult: initDetailedResult(),
        simpleSearch: initSimpleSearch(),
        detailedApplicationSearch: initApplicationsAdvancedSearch(),
        detailedFileSearch: initFilesDetailedSearch(),
        fileInfo: initFileInfo(),
        avatar: initAvatar(),
        filePreview: initFilePreview(),
        project: ProjectRedux.initialState,
        ...ApplicationRedux.init(),
        ...AccountingRedux.init(),
        ...FavoritesRedux.init(),
        ...SnackbarRedux.init(),
        responsive: undefined,
    }
}


export const initFilePreview = () => ({
    file: undefined,
    error: undefined
});

export type AvatarReduxObject = typeof defaultAvatar & { error?: string };
export const initAvatar = (): AvatarReduxObject => ({...defaultAvatar, error: undefined});

export const initSimpleSearch = (): SimpleSearchStateProps => ({
    files: emptyPage,
    filesLoading: false,
    applications: emptyPage,
    applicationsLoading: false,
    errors: [],
    search: "",
    applicationSearch: initApplicationsAdvancedSearch(),
    fileSearch: initFilesDetailedSearch()
});

export const initAnalyses = (): AnalysisReduxObject => ({
    page: emptyPage,
    loading: false,
    error: undefined,
    sortBy: RunsSortBy.createdAt,
    sortOrder: SortOrder.ASCENDING
});

export const initZenodo = (): ZenodoReduxObject => ({
    connected: false,
    loading: false,
    page: emptyPage
});

export const initSidebar = (): SidebarReduxObject => ({
    pp: false,
    kcCount: 0,
    options: []
});

export const initUploads = (): UploaderReduxObject => ({
    path: "",
    uploads: [],
    visible: false,
    allowMultiple: false,
    error: undefined,
    onFilesUploaded: () => null,
    loading: false
});

export const initFileInfo = (): FileInfoReduxObject => ({
    activity: emptyPage,
    loading: false
});

export const initFiles = (homeFolder: string): FilesReduxObject => ({
    page: emptyPage,
    sortOrder: getItemOrDefault("sortOrder", SortOrder.ASCENDING, SortOrder),
    sortBy: getItemOrDefault("sortBy", SortBy.PATH, SortBy),
    loading: false,
    error: undefined,
    path: "",
    filesInfoPath: "",
    sortingColumns: [getFilesSortingColumnOrDefault(0), getFilesSortingColumnOrDefault(1)],
    fileSelectorLoading: false,
    fileSelectorShown: false,
    fileSelectorPage: emptyPage,
    fileSelectorPath: homeFolder,
    fileSelectorIsFavorites: false,
    fileSelectorCallback: () => undefined,
    fileSelectorError: undefined,
    disallowedPaths: [],
    invalidPath: false
});

export const initFilesDetailedSearch = (): DetailedFileSearchReduxState => ({
    hidden: true,
    allowFolders: true,
    allowFiles: true,
    fileName: "",
    extensions: new Set(),
    tags: new Set(),
    sensitivities: new Set(),
    createdBefore: undefined,
    createdAfter: undefined,
    modifiedBefore: undefined,
    modifiedAfter: undefined,
    error: undefined,
    loading: false
});

export const initApplicationsAdvancedSearch = (): DetailedApplicationSearchReduxState => ({
    error: undefined,
    loading: false,
    hidden: true,
    appName: "",
    appVersion: "",
    tags: ""
});
