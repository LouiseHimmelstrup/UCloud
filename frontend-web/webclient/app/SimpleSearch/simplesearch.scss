html body div#app div div.container-wrapper div.container-content div.container-padding form.ui.form.form-input-margin {
    margin-bottom: 15px !important;
}