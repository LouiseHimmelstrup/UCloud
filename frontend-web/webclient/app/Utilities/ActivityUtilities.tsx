import { buildQueryString } from "./URIUtilities";

export const activityQuery = (offset: number | null, scrollSize: number) => {
    const params = { scrollSize };
    if (offset !== null) params["offset"] = offset;

    return buildQueryString("/activity/browse/user", params);
}

export const activityStreamByPath = (path: string) =>
    `/activity/by-path?path=${encodeURIComponent(path)}` 

export const getLatestActivity = () => undefined;