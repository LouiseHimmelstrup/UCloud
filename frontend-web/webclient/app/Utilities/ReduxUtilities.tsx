import dashboard from "Dashboard/Redux/DashboardReducer";
import {initObject} from "DefaultObjects";
import header, {CONTEXT_SWITCH, USER_LOGIN, USER_LOGOUT} from "Navigation/Redux/HeaderReducer";
import sidebar from "Navigation/Redux/SidebarReducer";
import status from "Navigation/Redux/StatusReducer";
import notifications from "Notifications/Redux/NotificationsReducer";
import * as ProjectRedux from "Project/Redux";
import {Action, AnyAction, combineReducers, createStore, Store} from "redux";
import {composeWithDevTools} from "redux-devtools-extension";
import {createResponsiveStateReducer, responsiveStoreEnhancer} from "redux-responsive";
import simpleSearch from "Search/Redux/SearchReducer";
import {responsiveBP} from "ui-components/theme";
import avatar from "UserSettings/Redux/AvataaarReducer";
import hookStore from "Utilities/ReduxHooks";

export function configureStore(
    initialObject: ReduxObject,
    reducers,
    enhancers?
): Store<ReduxObject, AnyAction> {
    const combinedReducers = combineReducers<ReduxObject, AnyAction>(reducers);
    const rootReducer = (state: ReduxObject, action: Action): ReduxObject => {
        if ([USER_LOGIN, USER_LOGOUT, CONTEXT_SWITCH].some(it => it === action.type)) {
            state = initObject();
        }
        return combinedReducers(state, action);
    };
    return createStore<ReduxObject, AnyAction, {}, {}>(rootReducer, initialObject, composeWithDevTools(enhancers));
}

export const responsive = createResponsiveStateReducer(
    responsiveBP,
    {infinity: "xxl"}
);

export const store = configureStore(initObject(), {
    dashboard,
    header,
    status,
    sidebar,
    notifications,
    simpleSearch,
    hookStore,
    avatar,
    loading,
    project: ProjectRedux.reducer,
    responsive: createResponsiveStateReducer(
        responsiveBP,
        {infinity: "xxl"}),
}, responsiveStoreEnhancer);

function loading(state = false, action: {type: string}): boolean {
    switch (action.type) {
        case "LOADING_START":
            return true;
        case "LOADING_END":
            return false;
        default:
            return state;
    }
}
