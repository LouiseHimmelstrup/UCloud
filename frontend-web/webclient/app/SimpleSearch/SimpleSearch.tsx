import * as React from "react";
import { Tab, List, Icon, Card, Message, Responsive, Form } from "semantic-ui-react";
import * as Pagination from "Pagination";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import { Cloud } from "Authentication/SDUCloudObject";
import * as UF from "UtilityFunctions";
import PromiseKeeper from "PromiseKeeper";
import { SingleApplication } from "Applications/Applications";
import { simpleSearch } from "Metadata/api";
import { SearchItem } from "Metadata/Search";
import { emptyPage } from "DefaultObjects";
import { AllFileOperations, getParentPath, getFilenameFromPath } from "Utilities/FileUtilities";
import { SimpleSearchProps, SimpleSearchState } from ".";
import { HeaderSearchType } from "DefaultObjects";
import { setPrioritizedSearch } from "Navigation/Redux/HeaderActions";
import { Application } from "Applications";
import { Page } from "Types";

class SimpleSearch extends React.Component<SimpleSearchProps, SimpleSearchState> {
    constructor(props) {
        super(props);
        this.state = {
            promises: new PromiseKeeper(),
            files: emptyPage,
            filesLoading: false,
            applications: emptyPage,
            applicationsLoading: false,
            projects: emptyPage,
            projectsLoading: false,
            error: "",
            search: this.props.match.params[0]
        };
    }

    componentDidMount() {
        if (!this.state.search) return;
        this.fetchAll(this.state.search);
    }

    shouldComponentUpdate(nextProps, _nextState): boolean {
        if (nextProps.match.params[0] !== this.props.match.params[0]) {
            this.setState(() => ({ search: nextProps.match.params[0] }));
            this.fetchAll(nextProps.match.params[0]);
        }
        if (nextProps.match.params.priority !== this.props.match.params.priority) {
            this.props.dispatch(setPrioritizedSearch(nextProps.match.params.priority as HeaderSearchType));
        }
        return true;
    }

    searchFiles = (search: string, pageNumber: number, itemsPerPage: number) => {
        const { promises } = this.state;
        this.setState(() => ({ filesLoading: true }));
        promises.makeCancelable(Cloud.get(`/file-search?query=${search}&page=${pageNumber}&itemsPerPage=${itemsPerPage}`))
            .promise.then(({ response }) => this.setState(() => ({ files: response })))
            .catch(_ => this.setState(() => ({ error: `${this.state.error}An error occurred searching for files. ` })))
            .finally(() => this.setState(() => ({ filesLoading: false })));
    }

    searchApplications = (search: string, pageNumber: number, itemsPerPage: number) => {
        const { promises } = this.state;
        this.setState(() => ({ applicationsLoading: true }));
        promises.makeCancelable(Cloud.get(`/hpc/apps?page=${pageNumber}&itemsPerPage=${itemsPerPage}`))
            .promise.then(({ response }: { response: Page<Application> }) => {
                response.items = response.items.filter(it => it.description.title.toLocaleLowerCase().includes(search.toLocaleLowerCase()));
                this.setState(() => ({ applications: response }))
            })
            .catch(() => this.setState(() => ({ error: `${this.state.error}An error occurred searching for applications. ` })))
            .finally(() => this.setState(() => ({ applicationsLoading: false })));
    }

    searchProjects = (search: string, pageNumber: number, itemsPerPage: number) => {
        const { promises } = this.state;
        this.setState(() => ({ projectsLoading: true }));
        promises.makeCancelable(simpleSearch(search, pageNumber, itemsPerPage))
            .promise.then(response => this.setState(() => ({ projects: response })))
            .catch(() => this.setState(() => ({ error: `${this.state.error}An error occurred searching for projects. ` })))
            .finally(() => this.setState(() => ({ projectsLoading: false })));
    }

    setPath = (t) => {
        const { text } = t;
        this.props.dispatch(setPrioritizedSearch(text as HeaderSearchType));
        this.props.history.push(`/simplesearch/${text.toLocaleLowerCase()}/${this.state.search}`);
    }

    fetchAll(search: string) {
        this.searchFiles(search, this.state.files.pageNumber, this.state.files.itemsPerPage);
        this.searchApplications(search, this.state.applications.pageNumber, this.state.applications.itemsPerPage);
        this.searchProjects(search, this.state.projects.pageNumber, this.state.projects.itemsPerPage);
    }

    search() {
        if (!this.state.search) return;
        this.props.history.push(`/simplesearch/${this.props.match.params.priority}/${this.state.search}`);
    }

    render() {
        const { search, files, projects, applications, filesLoading, applicationsLoading, projectsLoading, error } = this.state;
        const errorMessage = !!error ? (<Message color="red" content={error} onDismiss={() => this.setState(() => ({ error: "" }))} />) : null;
        // Currently missing ACLS to allow for fileOperations
        const fileOperations = AllFileOperations(true, false, () => this.searchFiles(search, files.pageNumber, files.itemsPerPage), this.props.history);
        const panes = [
            {
                menuItem: "Files", render: () => (
                    <Tab.Pane loading={filesLoading}>
                        <Pagination.List
                            loading={filesLoading}
                            pageRenderer={(page) => (<SimpleFileList files={page.items} />)}
                            page={files}
                            onItemsPerPageChanged={(itemsPerPage: number) => this.searchFiles(search, 0, itemsPerPage)}
                            onPageChanged={(pageNumber: number) => this.searchFiles(search, pageNumber, files.itemsPerPage)}
                        />
                    </Tab.Pane>)
            },
            {
                menuItem: "Projects", render: () => (
                    <Tab.Pane loading={projectsLoading}>
                        <Pagination.List
                            loading={projectsLoading}
                            pageRenderer={(page) => page.items.map((it, i) => (<SearchItem key={i} item={it} />))}
                            page={projects}
                            onItemsPerPageChanged={(itemsPerPage: number) => this.searchProjects(search, 0, itemsPerPage)}
                            onPageChanged={(pageNumber: number) => this.searchProjects(search, pageNumber, projects.itemsPerPage)}
                        />
                    </Tab.Pane>
                )
            },
            {
                menuItem: "Applications", render: () => (
                    <Tab.Pane loading={applicationsLoading}>
                        <Pagination.List
                            loading={applicationsLoading}
                            pageRenderer={({ items }) =>
                                <Card.Group className="card-margin">
                                    {items.map((app, i) => <SingleApplication key={i} app={app} />)}
                                </Card.Group>
                            }
                            page={applications}
                            onItemsPerPageChanged={(itemsPerPage: number) => this.searchApplications(search, 0, itemsPerPage)}
                            onPageChanged={(pageNumber: number) => this.searchApplications(search, pageNumber, applications.itemsPerPage)}
                        />
                    </Tab.Pane>
                )
            }
        ];
        return (
            <React.StrictMode>
                {errorMessage}
                <Responsive maxWidth={999} as={Form} className="form-input-margin" onSubmit={() => this.search()}>
                    <Form.Input style={{ marginBottom: "15px" }} onChange={(_, { value }) => this.setState(() => ({ search: value }))} fluid />
                </Responsive>
                <Tab onTabChange={({ target }) => this.setPath(target)} activeIndex={SearchPriorityToNumber(this.props.match.params.priority)} panes={panes} />
            </React.StrictMode>
        );
    }
};

const SimpleFileList = ({ files }) => (
    <List size="large" relaxed>
        {files.map((f, i) => (
            <List.Item key={i}>
                <List.Content>
                    <Icon name={UF.iconFromFilePath(f.path, f.fileType, Cloud.homeFolder)} size={undefined} color={"blue"} />
                    <Link to={`/files/${f.fileType === "FILE" ? getParentPath(f.path) : f.path}`}>
                        {getFilenameFromPath(f.path)}
                    </Link>
                </List.Content>
                {/* <FileOperations fileOperations={fileOperations} files={[f]} /> */}
                <List.Content />
            </List.Item>
        ))}
    </List>
);

const SearchPriorityToNumber = (search: String): number => {
    if (search.toLocaleLowerCase() === "projects") return 1;
    if (search.toLocaleLowerCase() === "applications") return 2;
    return 0;
}

export default connect()(SimpleSearch)