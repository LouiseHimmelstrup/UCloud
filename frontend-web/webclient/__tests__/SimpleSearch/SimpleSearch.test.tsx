import * as React from "react";
import SimpleSearch from "SimpleSearch/SimpleSearch";
import { create } from "react-test-renderer";
import { configureStore } from "Utilities/ReduxUtilities";
import { Provider } from "react-redux"
import { initNotifications } from "DefaultObjects";
import notifications from "Notifications/Redux/NotificationsReducer";

describe("Simple Search", () => {
    test("Mount simplesearch", () => {
        const store = configureStore({ notifications: initNotifications() }, { notifications });
        expect(create(
            <Provider store={store}>
                <SimpleSearch match={{ params: { 0: "", priority: "projects" } }} />
            </Provider>
        )).toMatchSnapshot();
    })
});