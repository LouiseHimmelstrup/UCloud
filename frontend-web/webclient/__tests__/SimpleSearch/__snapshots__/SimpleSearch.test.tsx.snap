// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Simple Search Mount simplesearch 1`] = `
<div>
  <div
    className="ui attached tabular menu"
  >
    <a
      className="item"
      onClick={[Function]}
    >
      Files
    </a>
    <a
      className="active item"
      onClick={[Function]}
    >
      Projects
    </a>
    <a
      className="item"
      onClick={[Function]}
    >
      Applications
    </a>
  </div>
  <div
    className="ui bottom attached segment active tab"
  >
    <span
      className="items-per-page-padding pagination-float-right"
    >
      <span>
        Items per page 
      </span>
      <div
        aria-expanded={false}
        className="ui compact inline dropdown"
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        role="listbox"
        tabIndex={0}
      >
        <div
          aria-live="polite"
          className="text"
          role="alert"
        >
          25
        </div>
        <i
          aria-hidden="true"
          className="dropdown icon"
          onClick={[Function]}
        />
        <div
          className="menu transition"
        >
          <div
            aria-checked={false}
            aria-selected={false}
            className="item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              10
            </span>
          </div>
          <div
            aria-checked={true}
            aria-selected={true}
            className="active selected item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              25
            </span>
          </div>
          <div
            aria-checked={false}
            aria-selected={false}
            className="item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              50
            </span>
          </div>
          <div
            aria-checked={false}
            aria-selected={false}
            className="item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              100
            </span>
          </div>
        </div>
      </div>
    </span>
    <div>
      <h2
        className="ui header"
      >
        No results.
        <a
          href="#"
          onClick={[Function]}
        >
           Try again?
        </a>
      </h2>
    </div>
    <div />
  </div>
</div>
`;
